cloudbridge/cloud/providers/azure/services.py

```diff
@@ -0,0 +1,1024 @@
+
+    def find(self, name, limit=None, marker=None):
+        """
+             Searches for a snapshot by a given list of attributes.
+        """
+        filters = {'Name': name}
+        cb_snapshots = [AzureSnapshot(self.provider, snapshot)
+                        for snapshot in azure_helpers.filter(
+                self.provider.azure_client.list_snapshots(), filters)]
+        return ClientPagedResultList(self.provider, cb_snapshots,
+                                     limit=limit, marker=marker)
+
+    def list(self, limit=None, marker=None):
+        """
+               List all snapshots.
+        """
+        snaps = [AzureSnapshot(self.provider, obj)
+                 for obj in
+                 self.provider.azure_client.list_snapshots()]
+        return ClientPagedResultList(self.provider, snaps, limit, marker)
+
+    def create(self, name, volume, description=None):
+        """
+        Creates a new snapshot of a given volume.
+        """
+        volume = (self.provider.block_store.volumes.get(volume)
+                  if isinstance(volume, str) else volume)
+
+        tags = {'Name': name}
+        snapshot_name = "{0}-{1}".format(name, uuid.uuid4().hex[:6])
+
+        if description:
+            tags.update(Description=description)
+
+        params = {
+            'location': self.provider.azure_client.region_name,
+            'creation_data': {
+                'create_option': 'Copy',
+                'source_uri': volume.resource_id
+            },
+            'disk_size_gb': volume.size,
+            'tags': tags
+        }
+
+        self.provider.azure_client. \
+            create_snapshot(snapshot_name, params)
+        azure_snap = self.provider.azure_client.get_snapshot(snapshot_name)
+        cb_snap = AzureSnapshot(self.provider, azure_snap)
+
+        return cb_snap
+
+
+class AzureComputeService(BaseComputeService):
+    def __init__(self, provider):
+        super(AzureComputeService, self).__init__(provider)
+        self._instance_type_svc = AzureInstanceTypesService(self.provider)
+        self._instance_svc = AzureInstanceService(self.provider)
+        self._region_svc = AzureRegionService(self.provider)
+        self._images_svc = AzureImageService(self.provider)
+
+    @property
+    def images(self):
+        return self._images_svc
+
+    @property
+    def instance_types(self):
+        return self._instance_type_svc
+
+    @property
+    def instances(self):
+        return self._instance_svc
+
+    @property
+    def regions(self):
+        return self._region_svc
+
+
+class AzureInstanceService(BaseInstanceService):
+    def __init__(self, provider):
+        super(AzureInstanceService, self).__init__(provider)
+
+    def create(self, name, image, instance_type, subnet=None, zone=None,
+               key_pair=None, security_groups=None, user_data=None,
+               launch_config=None, **kwargs):
+
+        instance_name = "{0}-{1}".format(name, uuid.uuid4().hex[:6])
```

Comment:
@madhugilla @afgane  I'm not too comfortable with the idea of truncating the first 6 characters, since that could make the uuid non-unique? Instead, we could enforce a length limit for a name at the cloudbridge interface level, and enforce that for all providers. What if we enforce RFC1035, which GCE also requires? This enforces some character constraints and limits the length to 64 characters.